---
title: "RNA-Seq Analysis"
author: "Syahirah"
date: "06/25/2021"
output: 
  html_document: 
    toc: true
    toc_depth: 2
    toc_float: true
bibliography: References.bib
link_citations: true
csl: cell.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warnings = FALSE,
  message = FALSE,
  fig.align = 'center')

# List of packages

listOfBiocPackages = c("DESeq2", "recount", "EnhancedVolcano","EnsDb.Hsapiens.v86", "msigdbr", "clusterProfiler" )
notInstalled <- which( !listOfBiocPackages %in% rownames(installed.packages()) )

## check there's still something left to install
if( length(notInstalled) ) {
    biocLite(listOfBiocPackages[ notInstalled ])
}

# Normal install
packages = c("knitr", "tidyverse", "ggpubr", "ggplot2", "pheatmap", "DT")

## Now load or install & load all
package.check <- lapply(
  packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE)
      library(x, character.only = TRUE)
    }
  }
)

#####

exp <-  'SRP043043'

url <- download_study(exp)

# Load data
load(file.path(exp, 'rse_gene.Rdata'))

# Add condition status
rse_gene$condition <- c(rep("siSCR", 3), rep("siZNF217", 3))

# Modify 'title' to indicate Replicates
rse_gene$cell = gsub('^[[:alpha:]]*[0-9]*\\_', '', rse_gene$title)

# Remove version from ensembl-id
rownames(rse_gene) <- gsub('\\.[0-9]*$', '', rownames(rse_gene))

# Create DEseq dataset: design formula variable of interest is treatment status
dds <- DESeqDataSet(rse_gene, ~condition)
dds$condition <- relevel(dds$condition, ref = "siSCR")

# Analyse
dds <- DESeq(dds)

# Obtain results, contrast to indicate which is ref.
res <- results(dds, contrast = c("condition", "siZNF217", "siSCR"))

# LFC shrink
normRes <- lfcShrink(dds, res = res, coef= "condition_siZNF217_vs_siSCR", type= "apeglm")

# Create PCA to inspect batch effect
# PTransform using vst
vsd<- vst(dds)

```

# Introduction

## ZNF217 in Breast Cancer

ZNF217 is a krupple-like is an oncogenic protein found in a chromosomal region frequently amplified in various cancers including breast cancer (Cohen et al., 2015). It is a known biomarker of poor prognosis and is found in 20-30% primary human breast cancer cases (Littlepage et al., 2012).

It belongs to the krupple-like family of transcription fators and is a transcriptional regulator invovled in both repressor and activating transcription complexes (Cohen et al., 2015). Overexpression of ZNF217 drives tumorigenic behavior such as impaired differentiation, self-renewal, motility, metastasis and impaired cell death.

## Experimental Design

While it is a known biomarker, understanding the downstream network it regulates is still ongoing. This study seeks to identify its gene targets through siRNA knockdown of ZNF217 in Oestrogen Receptor Positive (ER+) HER2-MCF7 cells with paired scrambled control. The experiment was done in triplicates and then total RNA was sequenced. 

# Analysis Workflow

## Identification of Differentially Expressed Genes (DEGs)
The dataset was obtained from recount3 (SRA ascession number: `r exp`) and DEGs were identified through the R package, DEseq2. The cutoffs used were `log2 |fold change| >= 1` and `p-value <0.05`.

Following which, gene set enrichment analysis (GSEA) was performed with the clusterProfiler package in R. Top pathways were identified by the normalized enrichment score (NES). 


# Data Quality Check

## Principal Component Analysis

In PCA, a clustering method is applied to essentially check if the treatment condition is the main variable responsible for variation in the data and to check if our replicates within each condition is similar. From the figure below, when separated by their treatment condition on PC1, the samples notably clusters by treatment condition and tightly within their treatment groups. We can see that the treatment condition accounts for 93% (PC1) of the variation and there is minimal variation between technical replicates (5%; PC2) which is ideal. 

```{r PCA-plot, results = 'last', fig.cap = "Fig.1: PCA of samples"}
# Create PCA plot
pcaData <- plotPCA(vsd, intgroup = "condition", returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = condition)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + labs(color='Treatment')
```

# Differentially Expressed Genes (DEGs) Analysis

## Volcano Plot of DEGs

```{r res-gene-df, include = FALSE}

resdf <- normRes %>%
  data.frame() %>%
  rownames_to_column(var="ensembl_id") 

# Creating a df of ENSEMBL (GENEID), SYMBOL, GENENAME using EnsDb.Hsapiens.v86 db
anno <- AnnotationDbi::select(EnsDb.Hsapiens.v86,resdf$ensembl_id, 
                                        columns=c("GENEID", "SYMBOL"), 
                                        keytype="GENEID")

# left_join the resdf and the gene information
finRes <- resdf %>%
  left_join(anno, by = c("ensembl_id" ="GENEID"))

sigRes <- finRes[which(finRes$padj < 0.05 & abs(finRes$log2FoldChange) >= 1.5 & finRes$baseMean >= 20), ]

total_upreg <- as.numeric(count(sigRes))
no_upreg <- as.numeric(count(sigRes %>% dplyr::filter(log2FoldChange > 1.5)))
no_downreg <- as.numeric(count(sigRes %>% dplyr::filter(log2FoldChange < -1.5)))

```

A volcano plot is typically used to quickly visualize the relationship between the p-value (or statisitcal significance) and log2 fold change in expression at the gene level. Using abritrary parameters ie. fold-change and p-value cut offs, we can visualize the proportion of genes that are signifcant genes and are up or down regulated.

In the figure below, we can see that in there is more significantly upregulated genes than downregulated genes in the siZNF217 samples compared to paired control. In total, there were **`r total_upreg` DEGs**, with **`r no_upreg` upregualted** and **`r no_downreg` downregulated** genes.

```{r volplot, fig.cap = "Fig.2: Volcano Plot of Differentially Expressed Genes in Treated vs Untreated Samples"}
# Plot Volcano
EnhancedVolcano( finRes, lab = finRes$SYMBOL, 
                 x = 'log2FoldChange', y = 'padj',
                 xlim = c(-8, 8), title = 'Treated vs Untreated',
                 pCutoff = 1e-100, FCcutoff = 1.25, 
                 pointSize = 2.0, 
                 labSize = 3.0,
                 border = "full", borderWidth = 1.5, borderColour = "black", 
                 gridlines.major = FALSE, gridlines.minor = FALSE)

```

## Heatmap of all DEGs

Through a heatmap we can visualize the differences in gene expression between samples of different treatment conditions and observe for conistency within the treatment group. In the figure below, a significant number of genes are upregulated in the knockdown sample when compared to the control.

```{r DEG,results = 'last', fig.cap = "Fig.3: Heatmap of differentially expressed genes"}
# Heatmap of all DEGs
mat <- assay(vsd)  
idx <- sigRes$ensembl_id 
DEgenes <- mat[idx,]

annotation <- as.data.frame(colData(vsd)[, c("cell", "condition")])

pheatmap(DEgenes, scale = "row", show_rownames = FALSE, clustering_distance_rows = "correlation", annotation_col = annotation)

```

## Top 20 Over- and Under-expressed DEGs

We then looked at the top 20 over and under expressed genes, visuazlied through a heatmap. Interestingly, the genes that were overexpressed are involved in **keratonicyte differentiation**, **epidermal differentiation**, and **cell signalling**. In particular, ***SCEL*** is invovled in terminal keratinocyte differentiation, ***FLG*** is involved in epidermal differentiation, ***ATP10D*** is involved in transport pathways, ***PDE7B*** and ***TLR1*** is involved in cell sginalling [3-7].

Interestingly, ***PART1*** a lncRNA reported to promote breast cancer progression through promoting proliferative and metastatic behaviors, was also upregulated.

Gene functions were retrieved from the GeneCards suite (Stelzer et al., 2016).

```{r Top20, results = 'last', fig.cap = c("Fig.4: Top 20 Over Expressed Genes", "Fig.5: Top 20 Under Expressed Genes")}
# Extract Top 20 Over and Under-expressed Genes
# Up
top20_up <- sigRes %>%
  dplyr::filter(log2FoldChange > 1.5 & padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  head(20)

# Down
top20_down <- sigRes %>%
  dplyr::filter(log2FoldChange < 1.5 & padj < 0.05) %>%
  arrange(log2FoldChange) %>%
 head(20)

# Plot Heatmaps
# Up HM
top20_up_hm <- mat[top20_up$ensembl_id,]
rownames(top20_up_hm) <- top20_up$SYMBOL

pheatmap(top20_up_hm, scale = "row", clustering_distance_rows = "correlation", annotation_col = annotation)

# Under HM
top20_down_hm <- mat[top20_down$ensembl_id,]
rownames(top20_down_hm) <- top20_down$SYMBOL

pheatmap(top20_down_hm, scale = "row", clustering_distance_rows = "correlation", annotation_col = annotation)

```

# Pathway Analysis
## GSEA

Since small changes in a single gene expression can have cascading effects, it would be relevant to look at changes in entire pathways to provide a more holistic view. This is done through GSEA analysis.
```{r GSEA, results = 'last'}
# Create gene set for ref
gene_sets <- msigdbr(species = "Homo sapiens", category = "C5")
gene_sets <- gene_sets %>%
  dplyr::select(gs_name, gene_symbol)

# Adding a score for GSEA
gsea_df <- finRes%>%
  arrange(padj) %>%
  mutate(gsea_metric = -log10(padj) * sign(log2FoldChange))

# Deal with inf (swap up the inf with the highest pvalue)
gsea_df <- gsea_df %>%
  mutate(padj = case_when(padj == 0 ~ .Machine$double.xmin,
                          TRUE ~ padj)) %>%
  mutate(gsea_metric = -log10(padj) * sign(log2FoldChange)) 

# Remove NAs and order by GSEA
gsea_df <- gsea_df  %>%
  dplyr::filter(!is.na(gsea_metric)) %>%
  arrange(desc(gsea_metric)) # needed to run GSEA later

# Get the ranked GSEA vector
ranks <- gsea_df  %>%
  dplyr::select(SYMBOL, gsea_metric) %>%
  distinct(SYMBOL, .keep_all = TRUE) %>%
  deframe()

# Run GSEA
set.seed(1234)
gseares <- GSEA(geneList = ranks, 
                TERM2GENE = gene_sets,
                seed = TRUE, 
                nPerm = 10000,
                pvalueCutoff = 0.25)

gsearesdf <- as.data.frame(gseares)

gsearesdt <- gsearesdf %>%
  dplyr::select("Description", "setSize", "enrichmentScore", "NES", "p.adjust", "qvalues")

datatable(gsearesdt, width = '100%', rownames = FALSE, caption = "Table 1: List of pathways enriched. Top enriched pathways were selected by NES.",
          options = list(
  autoWidth = TRUE,
  pageLength = 5,
  lengthMenu = c(5, 10, 15, 20),
  scrollX = TRUE
))
```

## Top Enriched Pathways

The gene-set used was Gene Ontology from MSigDB. The top 5 over expressed pathways enriched were identified by the highest NES scores while the top 5 under expressed pathways were identified by the lowest NES scores. The p-value cut-off for significance was 0.25. The top overexpressed pathways were consistent with the top over expressed DEGs. 

```{r top-gsea, include = FALSE}
# Plot GSEA results
# Top 5 Over
# Make GSEA plot for top and bottom results
top_pathways <- gsearesdf %>%
  top_n(n = 5, wt = NES) %>%
  pull(ID)

#  Make gseaplot for each and return as list
top_pathway_plots <- lapply(top_pathways, function(pathway) {
  gseaplot(gseares, geneSetID = pathway, title = pathway)
})

# Arrange with labels as a multi-panel plot
top_pathway_plot <- ggarrange(plotlist = top_pathway_plots,
                              ncol = 2, nrow = 3, labels = "AUTO")


#  Save it
ggsave(top_pathway_plot, filename = "top_GSEA_up.png",
       height = 11, width = 18)

# Top 5 Under
bottom_pathways <- gsearesdf %>%
  top_n(n = 5, wt = -NES) %>%
  pull(ID)

bottom_pathway_plots <- lapply(bottom_pathways, function(pathway) {
  gseaplot(gseares, geneSetID = pathway, title = pathway)
})

bottom_pathway_plot <- ggarrange(plotlist = bottom_pathway_plots,
                                 ncol = 2, nrow = 3, labels = "AUTO")

ggsave(bottom_pathway_plot, filename = "top_GSEA_down.png",
       height = 11, width = 18)
```

```{r gsea-figs, fig.cap = c("Fig.6: Top 5 Over Expressed Pathways", "Fig.7: Top 5 Under Expressed Pathways")}

include_graphics("top_GSEA_up.png")

include_graphics("top_GSEA_down.png")

```


# Discussion

Overexpression of genes and pathways relating to keratinocyte and epidermal differentiation in siZNF217 compared to control is unsurprising since ZNF217 is known to suppress cell differentiation. Additionally, GSEA analysis also showed overexpression of cornification pathway in the knockdown cells, which is programmed cell death specifically in the epidermis. The results suggest strongly that ZNF217 contributes to tumorigenesis by suppressing important cell differentiation and programmed cell death processes.

Additionally, underexpression of genes and pathways relating to cell proliferation suggest that ZNF217 promnoted tumour growth by way of proliferation.

Moving forward, selected genes should be validated using RT-qPCR. Genes promoting epidermal/keratinocyte differentiation could be possible therapeutic targets and can be further investigated via *in vitro* assays.

```{r appendix,  figures-side, fig.show="hold", out.width="50%",  fig.cap = c("Supp. Fig. 1: MA Plot of normalized counts against transformed log fold change.", "Supp. Fig. 2: Histogram of GSEA-matric.")}
# MA Plot
#plotMA(normRes)

# GSEA value histogram
#hist(gsea_df$gsea_metric, breaks = 100)
```

# References





